src/rgw/rgw_cache.h

```diff
@@ -160,7 +160,7 @@ class ObjectCache {
   unsigned long lru_size;
   unsigned long lru_counter;
   unsigned long lru_window;
-  ceph::shared_mutex lock = ceph::make_shared_mutex("ObjectCache");
+  ceph::mutex lock = ceph::make_mutex("ObjectCache");
```

Comment:
i'd expect cache reads to be far more common than the writes, so i'm skeptical about this change which prevents multiple readers from sharing this lock

Reply:
@cbodley, it's been a while since I was working on this so I will need to go back and see if I can dig out the performance data.  In the mean time, let me explain my rationale.  It's more expensive to lock std::shared_mutex than std::mutex due to the extra complexity involved in incrementing/decrementing the atomic counter of readers during lock/unlock.  IE the more often you are locking/unlocking, the worse this gets vs just using a simple mutex.  Some background from Anthony Williams on the topic:

https://lists.boost.org/Archives/boost/2010/11/173495.php

More recent testing in c++:
https://www.codeproject.com/Articles/1183423/We-Make-a-std-shared-mutex-10-Times-Faster

And C#:
http://joeduffyblog.com/2009/02/11/readerwriter-locks-and-their-lack-of-applicability-to-finegrained-synchronization/

That last example isn't super interesting for c++ std:shared_mutex specifically, but showcases the same kind of general behavior where you need both a low number of writers (~10%) and a high number of function calls to really see any benefit.  That seems to more or less agree with what was shown in the c++ article.  So for us, I think it really comes down to how often we are writing and how long we are holding those locks for.  I think it will need to both be very skewed toward readers and we must hold the locks for extended periods of time to justify the extra complexity in shared_mutex.

Reply:
Ok, I went back and looked through the test data I have and ended up remembering that the whole analysis got interrupted when I discovered fairly significant performance/cpu usage regressions in RGW between nautilus and pacific.  That data is here:

https://docs.google.com/spreadsheets/d/18WqVGdO_JxVWyboktK6N6raqSJZWBEyN2F4QOg29cqA/edit#gid=862600772

My thought at the time was that those issues are dominating the effect of the cache.  I think this PR is still worth looking at, but the much bigger issues appear to be when #31580 and #35335 were introduced.

Reply:
thanks Mark. i discussed this with @mkogan1 last week, and suggested that we separate these two commits for testing so we can verify this independently

Reply:
> thanks Mark. i discussed this with @mkogan1 last week, and suggested that we separate these two commits for testing so we can verify this independently

testing the performance of commit cb871801355adf9708eed1079b7cc74536e8beba cherry picked over current master with hsbench small object put workload, 
I see on average about **1.5 percent performance improvement** consistently over repeated testing iterations (after accounting for sampling noise)

Reply:
although the benchmarks results are fluctuating saw up to 3.8% performance increase with both commits
and `--rgw_cache_shard_count=37` on a 40 HT cores numa node. (with the default `--rgw_cache_shard_count=4` was not able to discern a performance difference)